### PR TITLE
ADR for API object updates

### PR DESCRIPTION
**Pull request by ntwyman**
## Description

Explain a little about the changes at a high level.

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?

## Verification Steps

* [x] All tests pass.
* [n/a] The requirements listed in [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely) have been satisfied.
* [n/a] There are no [aXe](https://www.deque.com/products/aXe/) warnings for UI.
* [n/a] This works in IE.

## References

* [Produce straw man TSP API documentation/workflow based docs sent to TSPs](https://www.pivotaltracker.com/story/show/155009577) for this change
* [Conversation in slack](https://trussworks.slack.com/archives/C0CUTR6P3/p1519853138000636) explains more about the approach used.

## Screenshots

N/A

### CODE REVIEW

File: docs/adr/0013-rest-api-updates.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+* Updating the Actual Delivery Date on a shipment, i.e. a TSP delivers the shipment and wants to record the delivery date
+* Accepting an Awarded Shipment, i.e. as shipment is awarded to a TSP and an agent for the TSP wants to accept that
+ shipment. In this case the client may not (or possibly cannot) exactly understand what state in the object should change
+ but clearly knows the action (`accept`) that they wish to perform.
+
+## Considered Alternatives
+
+* **Allowing state change via POST/PUT passing partial objects**, e.g.
+
+```http request
```
Comment (jim) [inline docs/adr/0013-rest-api-updates.md]: I think code blocks, etc. within these bulleted items need to be indented by 4 spaces in order to render properly.

File: docs/adr/0013-rest-api-updates.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+`PUT` is generally used to update or replace an object, i.e. `PUT /widgets/15435` would update or replace Widget 15435
+with the new value in the body of the request.
+
+Where the consensus breaks down is when a request wants to update or change just part of the state of an object, e.g.
+change the description of Widget 15435 to be, say, "Recently updated - now in blue". Here there is are a range of patterns
+ in play, some being easy to implement but technically ambiguous/confusing to those that are 'correct' as proscribed in
+  RFCs but in practice ignored and putting a greater burden on the clients of the API.
+
+In particular, we want to consider a couple of use cases, viz:
```
Comment (jim) [inline docs/adr/0013-rest-api-updates.md]: 💯 for including these specific examples.

File: docs/adr/0013-rest-api-updates.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+```http request
+POST /shipments/1243/accept HTTP/1.1
+Content-Type: application/json
+
+{
+  "reason": "Can accommodate this move"
+}
+```
+
+NOTE: This is not explicitly setting either an 'accept' nor a 'reason' property of a shipment, but can be thought of as an rpc, e.g.
```
Comment (jim) [inline docs/adr/0013-rest-api-updates.md]: This note really helps clarify how to think about these actions.

File: docs/adr/0013-rest-api-updates.md (hunk 1 above)
Comment (ntwyman) [reply docs/adr/0013-rest-api-updates.md]: True - fixed

**Comment by ntwyman** [comment]
@tinyels I am not sure I understand why this would make the client code have to change. Nothing precludes the client from using PUT and updating a whole object. 

The ADR is really written with clients of the public API in mind and for cases where, e.g. the TSP wants to accept and existing shipment or set the delivery date. For these cases I would expect custom code or a manually created form to be used.

I feel like I'm not understanding something about your comment.

File: docs/adr/0013-rest-api-updates.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+Whilst there is no official definition of the mapping used for a RESTful API there are some very common patterns like using
+ `POST /widgets` to create a new Widget, the value of the Widget being in the body of the request. Likewise `GET /widgets/15435`
+ will retrieve the Widget with ID 15435.
+
+`PUT` is generally used to update or replace an object, i.e. `PUT /widgets/15435` would update or replace Widget 15435
+with the new value in the body of the request.
+
+Where the consensus breaks down is when a request wants to update or change just part of the state of an object, e.g.
+change the description of Widget 15435 to be, say, "Recently updated - now in blue". Here there is are a range of patterns
```
Comment (sansb) [inline docs/adr/0013-rest-api-updates.md]: Here there ~is~ are a range of patterns

File: docs/adr/0013-rest-api-updates.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+ in play, some being easy to implement but technically ambiguous/confusing to those that are 'correct' as proscribed in
+  RFCs but in practice ignored and putting a greater burden on the clients of the API.
+
+In particular, we want to consider a couple of use cases, viz:
+
+* Updating the Actual Delivery Date on a shipment, i.e. a TSP delivers the shipment and wants to record the delivery date
+* Accepting an Awarded Shipment, i.e. as shipment is awarded to a TSP and an agent for the TSP wants to accept that
```
Comment (sansb) [inline docs/adr/0013-rest-api-updates.md]: Typo: as -> a

**Comment by tinyels** [comment]
Nick, 
I guess it wasn't clear to me that this was about the public api vs the internal api. Is that a distinction worth making?

**Comment by sansb** [comment]
@ntwyman LGTM. I also wasn't clear on this being specifically about the public API. If this was a decision concerning the private API as well, I would second @tinyels 's feedback about this needing to consider the client libraries we're currently using or planning to use since some client libraries are super opinionated about how they will interact with the API.

**Comment by ntwyman** [comment]
So this is not exclusively about the public API but is restricted to how the API CAN allow partial updates where they make sense. It doesn't say that the API or clients MUST support partial updates.

If the client tools are requiring that the API only publish endpoints that match a certain pattern then we should definitely talk about the opinions that the clients have. Again, I am not sure I am understanding the problem that is being raised. I'm gonna hold off merging this until Monday to give time to chat about this, but I think at this point we should get together on Zoom and make sure that there isn't something specific I'm overlooking

**Pull request merged**